qiskit/synthesis/unitary/qsd.py

```diff
@@ -221,56 +316,63 @@ def _demultiplex(um0, um1, opt_a1=False, opt_a2=False, *, _depth=0, _ctrl_index=
 
-    # multiplexed Rz
+    # multiplexed Rz gate
+    # If opt_a1 = ``True``, then we reduce 2 ``cx`` gates per call.
     angles = 2 * np.angle(np.conj(dvals))
-    ucrz = UCRZGate(angles.tolist())
+    if _vw_type == "only_w" and opt_a1:
+        ucrz = _get_ucrz(nqubits, angles)
+    elif _vw_type == "only_v" and opt_a1:
+        ucrz = _get_ucrz(nqubits, angles).reverse_ops()
+    else:
+        ucrz = _get_ucrz(nqubits, angles, _vw_type="all")
     circ.append(ucrz, [layout[-1]] + layout[: nqubits - 1])
 
-    # right gate
-    right_gate = qs_decomposition(
-        vmat, opt_a1=opt_a1, opt_a2=opt_a2, _depth=_depth + 1
-    ).to_instruction()
-    circ.append(right_gate, layout[: nqubits - 1])
+    # right gate. In this case we decompose vmat.
+    # Otherwise, it is combined with the B matrix.
+    if _vw_type in ["only_v", "all"]:
+        right_gate = qs_decomposition(
+            vmat, opt_a1=opt_a1, opt_a2=opt_a2, _depth=_depth + 1
+        ).to_instruction()
+        circ.append(right_gate, layout[: nqubits - 1])
 
-    return circ
+    return circ, vmat, wmat
 
 
-def _get_ucry_cz(nqubits, angles):
-    """
-    Get uniformly controlled Ry gate in CZ-Ry as in UCPauliRotGate.
-    """
-    nangles = len(angles)
-    qc = QuantumCircuit(nqubits)
-    q_controls = qc.qubits[:-1]
-    q_target = qc.qubits[-1]
-    if not q_controls:
-        if np.abs(angles[0]) > _EPS:
-            qc.ry(angles[0], q_target)
-    else:
-        angles = angles.copy()
-        UCPauliRotGate._dec_uc_rotations(angles, 0, len(angles), False)
-        for i, angle in enumerate(angles):
-            if np.abs(angle) > _EPS:
-                qc.ry(angle, q_target)
-            if not i == len(angles) - 1:
-                binary_rep = np.binary_repr(i + 1)
-                q_contr_index = len(binary_rep) - len(binary_rep.rstrip("0"))
-            else:
-                # Handle special case:
-                q_contr_index = len(q_controls) - 1
-            # leave off last CZ for merging with adjacent UCG
-            if i < nangles - 1:
-                qc.cz(q_controls[q_contr_index], q_target)
-    return qc
+def _get_ucrz(nqubits, angles, _vw_type=None):
+    """This function synthesizes UCRZ without the final CX gate,
+    unless _vw_type = ``all``."""
+    circuit = QuantumCircuit(nqubits)
+    q_controls = circuit.qubits[1:]
+    q_target = circuit.qubits[0]
+
+    UCPauliRotGate._dec_uc_rotations(angles, 0, len(angles), False)
+    for i, angle in enumerate(angles):
+        if np.abs(angle) > _EPS:
+            circuit.rz(angle, q_target)
+        if not i == len(angles) - 1:
+            binary_rep = np.binary_repr(i + 1)
+            q_contr_index = len(binary_rep) - len(binary_rep.rstrip("0"))
+            circuit.cx(q_controls[q_contr_index], q_target)
+        elif _vw_type == "all":
+            q_contr_index = len(q_controls) - 1
+            circuit.cx(q_controls[q_contr_index], q_target)
```

Comment:
why repeat this line twice in each branch instead of putting in once after the branch?